File: compiler/java/com/google/dart/compiler/type/InferredType.java
// Copyright (c) 2012, the Dart project authors. Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

package com.google.dart.compiler.type;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;
import java.lang.reflect.Proxy;

/**
 * Marker interface for {@link Type} which means that this {@link Type} was not specified by user,
 * but instead inferred from context.
 */
public interface InferredType {
  public static class Helper {
    /**
     * @return the mix of the given {@link Type} interface and {@link InferredType}.
     */
    public static Type make(final Type type) {
      if (type instanceof InterfaceType) {
        return (Type) Proxy.newProxyInstance(type.getClass().getClassLoader(), new Class<?>[] {
            InterfaceType.class,
            InferredType.class}, new InvocationHandler() {
          @Override
          public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
            return method.invoke(type, args);
          }
        });
      }
      return type;
    }
  }
}

File: pkg/front_end/testcases/primary_constructors/inferred_type.dart
// Copyright (c) 2026, the Dart project authors. Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

class C1(a, var b, final c) {
  int d = a; // Error
  int e = b; // Error
  int f = c; // Error
}

class C2([a = null, var b = null, final c = null]) {
  int d = a; // Error
  int e = b; // Error
  int f = c; // Error
}

class C3({a = null, var b = null, final c = null}) {
  int d = a; // Error
  int e = b; // Error
  int f = c; // Error
}

extension type ET1(a);

extension type ET2([a = null]);

extension type ET3({a = null});

test(C1 c1, C2 c2, C3 c3, ET1 e1, ET2 e2, ET3 e3) {
  int a = c1.b; // Error
  int b = c1.c; // Error
  int c = c2.b; // Error
  int d = c2.c; // Error
  int e = c3.b; // Error
  int f = c3.c; // Error
  int g = e1.a; // Error
  int h = e2.a; // Error
  int i = e3.a; // Error
}
